src/loaders/compute_tools/microtrait/microtrait_util.py

```diff
@@ -0,0 +1,59 @@
+from collections import defaultdict
+from pathlib import Path
+
+import pandas as pd
+
+from src.loaders.common.loader_common_names import SYS_TRAIT_ID
+
+# column names from the rule2trait file that contain the trait names
+_RULE2TRAIT_TRAIT_NAME_COLS = ['microtrait_trait-name1', 'microtrait_trait-name2', 'microtrait_trait-name3']
+_RULE2TRAIT_RULE_NAME_COL = 'microtrait_rule-name'  # column name from the rule2trait file that contains the rule name
+
+# `microtrait_ruleunwrapped.txt` file unfortunately has no header, so we need to specify the column names
+_UNWRAPPED_RULE_COL = 'unwrapped_rule'
+_RULEUNWRAPPED_RULE_NAME_COL = 'rule_name'
+_RULEUNWRAPPED_COLS = [
+    _RULEUNWRAPPED_RULE_NAME_COL,
+    'rule_boolean',
+    'rule_display_name',
+    _UNWRAPPED_RULE_COL]
+
+
+def create_trait_unwrapped_rules(
+        rule2trait_file: Path,
+        ruleunwrapped_file: Path,
+        trait_unwrapped_rules_file: Path = None,
+) -> dict[str, set[str]]:
+    """
+    Generate a mapping of traits to unwrapped rule expressions.
+    """
+    trait_rule_unwrapped_mapping = defaultdict(set)
+
+    # read the rule2trait file and populate rule names to each trait
+    rule2trait_df = pd.read_csv(rule2trait_file, sep='\t')
+    for index, row in rule2trait_df.iterrows():
+        rule_name = row[_RULE2TRAIT_RULE_NAME_COL]
+        for trait_name_col in _RULE2TRAIT_TRAIT_NAME_COLS:
+            trait_name = row[trait_name_col]
+            trait_rule_unwrapped_mapping[trait_name].add(rule_name) if pd.notna(trait_name) else None
+
+    # read the ruleunwrapped file and populate unwrapped rule expressions to each trait
+    ruleunwrapped_df = pd.read_csv(ruleunwrapped_file, sep='\t', names=_RULEUNWRAPPED_COLS)
+    for trait_name, rule_names in trait_rule_unwrapped_mapping.items():
+        for rule_name in rule_names.copy():
+            unwrapped_rule = ruleunwrapped_df.loc[
+                ruleunwrapped_df[_RULEUNWRAPPED_RULE_NAME_COL] == rule_name, _UNWRAPPED_RULE_COL].values
```

Comment:
IIUC, this is running through the pandas data frame for each rule_name and then pulling out the unwrapped rule, so this is O(rule_names^2). If that's right it'd be more efficient (and arguably easier to read) to read the unwrapped rules into a `dict` and pull them from there

Reply:
👍

Reply:
:+1: